### PR TITLE
ENS Name Resolution

### PR DESCRIPTION
**Pull request by Jon-edge**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

 none

### Description

 none

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205920007236365

### CODE REVIEW

File: src/common/utils.ts [hunk 1]
```diff
@@ -477,3 +483,52 @@ export const objectCheckOneWay = (obj1: any, obj2: any): boolean => {
...
+    rpcServer => async () => {
+      const ethProvider = new ethers.providers.JsonRpcProvider(
+        addRpcApiKey(initOptions, pluginId, rpcServer),
+        chainId
+      )
+      return await func(ethProvider)
+    }
+  )
+  return await asyncWaterfall(shuffleArray(funcs))
+}
```
Comment (samholmes) [inline src/common/utils.ts:534]: This should be in its own commit.

File: src/common/utils.ts [hunk 2]
```diff
@@ -141,7 +147,7 @@ export const snoozeReject: Function = async (ms: number) =>
 export const snooze: Function = async (ms: number) =>
   await new Promise((resolve: Function) => setTimeout(resolve, ms))
 
-export async function promiseAny<T>(promises: Array<Promise<T>>): Promise<T> {
+export async function promiseAny(promises: Array<Promise<any>>): Promise<any> {
```
Comment (samholmes) [inline src/common/utils.ts:150]: Regression.

File: src/ethereum/networkAdapters/RpcAdapter.ts [hunk 3]
```diff
@@ -396,7 +377,11 @@ export class RpcAdapter
       params
     }
 
-    url = this.addRpcApiKey(url)
+    url = addRpcApiKey(
+      this.ethEngine.initOptions,
+      this.ethEngine.currencyInfo.pluginId,
+      url
+    )
```
Comment (samholmes) [inline src/ethereum/networkAdapters/RpcAdapter.ts:384]: I think I've discovered this to be a bug. We're adding API key to the server urls in the constructor, so the url passed into this call should already include the api key. Make sure this private method is always called with the a `baseUrl` from the RpcAdapter config, and if so, remove this as a separate fix commit.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 ## Unreleased
 
-## 2.12.0 (2023-11-14)
```
Comment (samholmes) [inline CHANGELOG.md:5]: This should have been fixed because the version was published with some real changes. Removing it is not the right thing. I added a fix commit to master, but be sure to resolve your conflict here correctly.

File: src/ethereum/EthereumTools.ts [hunk 5]
```diff
@@ -14,15 +14,21 @@ import {
...
+import { ethers } from 'ethers'
 
 import { PluginEnvironment } from '../common/innerPlugin'
 import { asMaybeContractLocation, validateToken } from '../common/tokenHelpers'
 import { encodeUriCommon, parseUriCommon } from '../common/uriHelpers'
-import { biggyScience, getLegacyDenomination } from '../common/utils'
+import {
+  biggyScience,
+  getLegacyDenomination,
+  multicastEthProviders
```
Comment (samholmes) [inline src/ethereum/EthereumTools.ts:25]: This is the commit which introduces `multicastEthProviders`

File: src/common/utils.ts [hunk 6]
```diff
@@ -477,3 +483,52 @@ export const objectCheckOneWay = (obj1: any, obj2: any): boolean => {
...
+  rpcServers: string[]
+  initOptions: EthereumInitOptions
+  pluginId: string
+  chainId?: number
+}): Promise<any> => {
+  const { func, rpcServers, initOptions, pluginId, chainId } = props
+  const funcs: Array<() => Promise<any>> = rpcServers.map(
+    rpcServer => async () => {
+      const ethProvider = new ethers.providers.JsonRpcProvider(
+        addRpcApiKey(initOptions, pluginId, rpcServer),
```
Comment (samholmes) [inline src/common/utils.ts:527]: Add the API key before calling `multicastEthProviders` and now `multicastEthProviders` doesn't need the extra parameters.

Also, instead of `rpcServers: string[]` you can paramterize `providers: ethers.providers.JsonRpcProvider[]`. In the end the function signature could look like:

```ts
export const multicastEthProviders = async <R, P extends ethers.providers.Provider>(props: {
  func: (ethProvider: P) => Promise<R>
  providers: P[]
}): Promise<R>
```

If you want to be more narrow then `Providers` type then change the subtype for `P` to `ethers.providers.JsonRpcProvider`.

File: src/ethereum/EthereumTools.ts [hunk 7]
```diff
@@ -357,8 +365,35 @@ export class EthereumTools implements EdgeCurrencyTools {
     }
     return cleanLocation.contractAddress.toLowerCase().replace(/^0x/, '')
   }
+
+  // #region otherMethods
+
+  /**
+   * Resolve an ENS name, for example: "bob.eth"
+   */
+  async resolveEnsName(ensName: string): Promise<string> {
```
Comment (samholmes) [inline src/ethereum/EthereumTools.ts:374]: Why not make this called `resolveName(name: string): Promise<string>` just incase other name services are deployed besides ENS?

**Review by samholmes — CHANGES_REQUESTED** [review]

File: src/ethereum/EthereumTools.ts (hunk 7 above)
Comment (Jon-edge) [reply src/ethereum/EthereumTools.ts:374]: This implementation is specific to ENS since it only works for ENS on ETH. If we deploy other name services that can be generic (including support in ETH), then we can do this

Comment (samholmes) [reply src/ethereum/EthereumTools.ts:374]: Fair enough, though I'd imagine the function signature would be the same. I only thought that maybe if we make it generic and just fail on the input if it's not an ENS name (or whatever the implementation supports), then there wouldn't be a need for a new method later on. It's forward thinking, so not necessary.

File: src/ethereum/networkAdapters/RpcAdapter.ts (hunk 3 above)
Comment (Jon-edge) [reply src/ethereum/networkAdapters/RpcAdapter.ts:384]: No the constructor requires an `EthereumEngine`, so it won't work.

Comment (Jon-edge) [reply src/ethereum/networkAdapters/RpcAdapter.ts:384]: `addRpcApiKey `isn't needed. The api keys are always populated in the `networkAdapterConfig`

**Review by samholmes — APPROVED** [review]
